CONTRIBUTING.md

```diff
@@ -1,81 +1,126 @@
 
-## Version policy and development branches
+## 3 Version policy and development branches
 
 We basically develop in the `master` branch.
 
 1. We will keep the first version digit `0` until we have some super major changes in the project organization level.
 
-2. The second version digit will be updated when we have major updates including new functions and refactoring, and 
+2. The second version digit will be updated when we have major updates, including new functions and refactoring, and 
    their related bug fix and recipe changes.
-   This version update will be done roughly at every half year so far (but it depends on the development plan).
+   This version update will be done roughly every half year so far (but it depends on the development plan).
 
-3. The third version digit will be updated when we fix serious bugs or accumulate some minor changes including
+3. The third version digit will be updated when we fix serious bugs or accumulate some minor changes, including
    recipe related changes periodically (every two months or so).
 
-## Unit testing
+## 4 Unit testing
 
 ESPnet's testing is located under `test/`.  You can install additional packages for testing as follows:
 ``` console
 $ cd <espnet_root>
+$ . ./tools/activate_python.sh
 $ pip install -e ".[test]"
 ```
 
-### Python
+### 4.1 Python
 
 Then you can run the entire test suite using [flake8](http://flake8.pycqa.org/en/latest/), [autopep8](https://github.com/hhatto/autopep8), [black](https://github.com/psf/black) and [pytest](https://docs.pytest.org/en/latest/) with [coverage](https://pytest-cov.readthedocs.io/en/latest/reporting.html) by
 ``` console
 ./ci/test_python.sh
 ```
+Followings are some useful tips when you are using pytest:
+- To create a new test file. write functions named like `def test_yyy(...)` in files like `test_xxx.py` under `test/`.
+[Pytest](https://docs.pytest.org/en/latest/) will automatically test them.
+- Note that, [pytest-timeouts](https://pypi.org/project/pytest-timeouts/) raises **an error when any tests exceed 2.0 sec**. To keep unit tests fast, please avoid large parameters, dynamic imports, and file access. If your unit test really needs more time, you can annotate your test function with `@pytest.mark.timeout(sec)`.
+- You can find pytest fixtures in `test/conftest.py`. [They finalize unit tests.](https://docs.pytest.org/en/latest/fixture.html#using-fixtures-from-classes-modules-or-projects)
+- As it is important to make sure that the unit test covers more codes, we recommend you to use `pytest --cov-report term-missing --cov=<target_dir> tests/` to check the status of test coverage. For more details, please refer to [coverage-test](https://pytest-cov.readthedocs.io/en/latest/readme.html).
+- For unit tests, we recommend you to separate large tests (e.g., `test_e2e_xxx.py`) into smaller ones, each one should test only methods/operations inside one file when it's possible. 
+   
```

Comment:
- New test file should be put under `test/` directory and named `test_xxx.py`. Each methods in test file should have the format `def test_yyy(...)`.  [Pytest](https://docs.pytest.org/en/latest/) will automatically find and test them.
- We recommend adding several small test files instead of grouping them together in one big file (e.g.: test_e2e_xxx.py`). Technically, a test file should only cover methods from one file (e.g.: `test_transformer_utils.py` to test `transformer_utils.py`).
- To monitor test coverage and avoid overlapping test, we recommend using  `pytest --cov-report term-missing <test_file|dir>` to highlight covered and missed lines. For more details, please refer to [coverage-test](https://pytest-cov.readthedocs.io/en/latest/readme.html).
- We limited test running time to 2.0 seconds (see: [pytest-timeouts](https://pypi.org/project/pytest-timeouts/)). As such, we recommend using small model parameters and avoiding dynamic imports, file access and unnecessary loops. If a unit test needs more running time, you can annotate your test with `@pytest.mark.timeout(sec)`.
- For test initialization (parameters, modules, etc), you can use pytest fixtures. Refer to  [pytest fixtures](https://docs.pytest.org/en/latest/fixture.html#using-fixtures-from-classes-modules-or-projects) for more information.